File: lazarus/backup/makeDoc.pas
//D> <a id="makeDoc"></a>
//D> # makeDoc
//D>
//D> |               | Description                                                               |
//D> | ------------: | ------------------------------------------------------------------------- |
//D> |               | Lazarus/Free Pascal project to export documentation lines with tag marker |
//D> | **File**      | makeDoc.pas                                                               |
//D> | **Author**    | [Leandro](mailto:leandrohuff@email.com)                                   |
//D> | **Date**      | _2025-10-06_                                                              |
//D> | **Version**   | _1.0.0_ / _2025-10-06_                                                    |
//D> | **Copyright** | CC01 1.0 Universal                                                        |
//D> | **Copyright** | **Copyright Icons** [Uicons by Flaticon](https://www.flaticon.com/uicons) |
//D>
//D> **Note**: Changes in this document will be discarded on next build,  
//D> any changes should be made on source code documentation instead.  
//D>
//D> <a id="Details"></a>
//D> ## Details
//D>
//D> Save formatted lines from source code into documentation file.  
//D> Read source code line-by-line and save prefixed lines by tag ??D to file.  
//D> C/C++/Free Pascal source codes lines start with tag //D and Bash lines start with tag //D.  
//D> Only those lines started by tags are exportedd to documentation files.  
//D> Mixed commented lines can co-exist at same source code, one for local  
//D> documentation purpose and another to be exported to apropriate documentation file.  
//D> All lines are documented using Markdown format, the exported document can be read  
//D> by an Markdown program reader.  
//D>
//D> <a id="Index"></a>
//D> ## Index
//D>
//D> | Index                             | Description                                         |
//D> | :-------------------------------- | :-------------------------------------------------- |
//D> | [Top](#makeDoc)                   |                                                     |
//D> | [Details](#Details)               |                                                     |
//D> | [Unit](#Unit)                     |                                                     |
//D> | [Type](#Type)                     |                                                     |
//D> | [Interface](#Interface)           |                                                     |
//D> | [Uses](#Uses)                     |                                                     |
//D> | [Constants](#Constants)           |                                                     |
//D> | [Variables](#Variables)           |                                                     |
//D> | [Implementation](#Implementation) |                                                     |
//D> | [btnAboutClick](#btnAboutClick)   |                                                     |
//D> | [btnCloseClick](#btnCloseClick)   |                                                     |
//D> | [btnRunClick](#btnRunClick)       |                                                     |
//D> | [OnClick](#OnClick)               |                                                     |
//D> | [OnCreate](#OnCreate)             |                                                     |
//D> | [Bottom](#updateEditDestine)      |                                                     |
//D>
//D> [⬅️](#makeDoc) | [⬆️](#Index)
//D>
//D> <a id="Unit"></a>
//D> ## Unit

unit makeDoc;

{$mode objfpc}{$H+}

//D>
//D> <a id="Interface"></a>
//D> ## Interface

interface

//D>
//D> <a id="Uses"></a>
//D> ## Uses
//D>
//D> | Name     | Description |
//D> | -------: | ----------- |
//D> | LCLType  | Key codes   |
//D> | Classes  | Abstract and concrete classes for object-oriented programming |
//D> | Forms    | Creating and managing user interface Forms |
//D> | Controls | Common controls, such as TButton, TEdit, TMemo, TComboBox, and TListBox |
//D> | Graphics | Native graphics classes like TCanvas, TFont, and TBitmap |
//D> | Dialogs  | Standard dialog components and functions like ShowMessage and dialogs for file opening/saving, color selection, and more |
//D> | StdCtrls | Standard components like TButton, TLabel, TEdit, TMemo, TCheckBox, and TRadioButton |
//D> | EditBtn  | Edit box with an associated button, is found in the LCLBase unit |
//D> | SysUtils | Part of the Free Pascal Run-Time Library (RTL) |
//D> | RegExpr  | Working with regular expressions is RegExpr |
//D>
//D> [⬅️](#Details) | [⬆️](#Index)

uses
    LCLType, Classes, Forms, Controls, Graphics, Dialogs, StdCtrls, EditBtn,
    ComCtrls, SysUtils, RegExpr, Dos;

//D>
//D> <a id="Type"></a>
//D> ## Type
//D>
//D> | Type            | Name                  | Description                     |
//D> | --------------: | --------------------: | ------------------------------- |
//D> | _TFormMain_     | **FormMain**          | Main form application           |
//D> | _TButton_       | **btnClose**          | Button close                    |
//D> | _TButton_       | **btnAbout**          | Show about box                  |
//D> | _TButton_       | **btnRun**            | Execute documentation procedure |
//D> | _TFileNameEdit_ | **editSource**        | Source file edit box            |
//D> | _TFileNameEdit_ | **editDestine**       | Destine file edit box           |
//D> | _TLabel_        | **lblStatus**         | Status processing               |
//D> | _TLabel_        | **lblDestine**        | Source file static label        |
//D> | _TLabel_        | **lblSource**         | Destine file static label       |
//D> | _procedure_     | **btnAboutClick**     | Treat About button click event  |
//D> | _procedure_     | **btnCloseClick**     | Treat Close button click event  |
//D> | _procedure_     | **btnRunClick**       | Treat Run button click event    |
//D> | _procedure_     | **OnClick**           | Treat edit destine click event  |
//D> | _procedure_     | **OnCreate**          | Treat on create form event      |
//D>
//D> [⬅️](#Uses) | [⬆️](#Index)

type

{ TFormMain }

    TFormMain = class(TForm)
        btnClose: TButton;
        btnAbout: TButton;
        btnRun: TButton;
        editSource: TFileNameEdit;
        editDestine: TFileNameEdit;
        lblStatus: TLabel;
        lblDestine: TLabel;
        lblSource: TLabel;
        procedure btnAboutClick(Sender: TObject);
        procedure btnCloseClick( Sender: TObject);
        procedure btnRunClick( Sender: TObject);
        procedure OnClick( Sender: TObject);
        procedure OnCreate( Sender: TObject);
    private
        currentDir : string;
        docDir : string;
    public

end;

//D>
//D> <a id="Constants"></a>
//D> ## Constants
//D>
//D> | Type | Name | Description |
//D> | :--: | :--: | :---------: |
//D> |  -   |  -   |      -      |
//D>
//D> <a id="Variables"></a>
//D> ## Variables
//D>
//D> | Type      | Name     | Description           |
//D> | --------: | -------: | --------------------- |
//D> | TFormMain | FormMain | Main form application |
//D>
//D> [⬅️](#Type) | [⬆️](#Index)

var
    FormMain: TFormMain;

//D>
//D> <a id="Implementation"></a>
//D> ## Implementation

implementation

{$R *.lfm}

{ TFormMain }

//D>
//D> <a id="btnAboutClick"></a>
//D> ### btnAboutClick
//D>
//D> | Return | Name | Parameters | Result |
//D> | ------ | ---- | ---------- | ------ |
//D> | _boolean_ | **btnAboutClick** | _TObject_ **Sender** | _None_ |
//D>
//D> Treat About button click event.  
//D>
//D> **Parameter**:  
//D> _TObject_ **Sender**: Objects and Events from main window.  
//D>
//D> **Result**:  
//D> _None_: Create and show a formatted about form window.  
//D>
//D> **Return**:  
//D> _boolean_  
//D>
//D> [⬅️](#Variables) | [⬆️](#Index)

procedure TFormMain.btnAboutClick(Sender: TObject);
begin
    ShowMessage('Export Documentation from taged lines.' + sLineBreak +
                'File: makeDoc' + sLineBreak +
                'Version: 1.0.0 / 2025-10-05' + sLineBreak +
                'Author: Leandro' + sLineBreak +
                'Copyright: CC01  1.0 Universal');
end;

//D>
//D> <a id="btnCloseClick"></a>
//D> ### btnCloseClick
//D>
//D> | Return | Name | Parameters | Result |
//D> | ------ | ---- | ---------- | ------ |
//D> | _boolean_ | **btnCloseClick** | _TObject_ **Sender** | _None_ |
//D>
//D> Treat Close button click event.  
//D>
//D> **Parameter**:  
//D> _TObject_ **Sender** : Objects and Events from main window.  
//D>
//D> **Result**:  
//D> _TFormMain_ : End and close main form window.  
//D>
//D> **Return**:  
//D> _None_
//D>
//D> [⬅️](#btnAboutClick) | [⬆️](#Index)

procedure TFormMain.btnCloseClick( Sender: TObject);
begin
    Close;
end;

//D>
//D> <a id="btnRunClick"></a>
//D> ### btnRunClick
//D>
//D> | Return | Name | Parameters | Result |
//D> | ------ | ---- | ---------- | ------ |
//D> | _None_ | **btnRunClick** | _TObject_ **Sender** | _None_ |
//D>
//D> Treat Run button click event.  
//D>
//D> **Parameter**:  
//D> _TObject_ **Sender**: Objects and Events from main window.  
//D>
//D> **Result**:  
//D> _TFormMain_: Run documentation procedure.  
//D>
//D> **Return**:  
//D> _None_  
//D>
//D> [⬅️](#btnCloseClick) | [⬆️](#Index)

procedure TFormMain.btnRunClick( Sender: TObject);
var

    inputFile, outputFile : TextFile;
    line : string;
    strList : TStringList;
    enableDoc : boolean;
    reAll, reDoc, reBegin, reEnd, reValidFilename : TRegExpr;

begin
    lblStatus.Caption := 'Running...';
    reValidFilename := TRegExpr.Create('([A-Za-z0-9_]+.?[A-Za-z0-9_]*)$');
    if FileExists (editSource.Text) then
    begin
        if not reValidFilename.Exec(editSource.Text) then
        begin
            ShowMessage('Source filename is not valid.');
            reValidFilename.Free;
            Exit;
        end;
        if not FileExists(editSource.Text) then
        begin
            ShowMessage('Source filename not found.');
            reValidFilename.Free;
            Exit;
        end;
        if DirectoryExists(editSource.Text) then
        begin
            ShowMessage('Source filename is a directory.' + sLineBreak +
            'Select a file and try again.');
            reValidFilename.Free;
            Exit;
        end;
        if not reValidFilename.Exec(editDestine.Text) then
        begin
            ShowMessage('Destine filename is not valid.');
            reValidFilename.Free;
            Exit;
        end;
        if DirectoryExists(editDestine.Text) then
        begin
            ShowMessage('Destine filename is a directory.' + sLineBreak +
            'Select a file and try again.');
            reValidFilename.Free;
            Exit;
        end;
        if FileExists(editDestine.Text) then
            if MessageDlg('Output filename already exist.' + sLineBreak +
                          'Do you LET OVERWRITE it?',
                          mtWarning,
                          [mbYes,
                          mbCancel],
                          0) = mrCancel then
            begin
                reValidFilename.Free;
                Exit;
            end;
        FormMain.Update;
        try
            AssignFile(inputFile, editSource.Text);
            Reset(inputFile);
            AssignFile(outputFile, editDestine.Text);
            Rewrite(outputFile);
            reAll    := TRegExpr.Create('^ *(#+|[/%-]{2,}[DBE])>? *');
            reDoc    := TRegExpr.Create('^ *(#+|[/%-]{2,}D)>? *');
            reBegin  := TRegExpr.Create('^ *(#+|[/%-]{2,}B)>? *');
            reEnd    := TRegExpr.Create('^ *(#+|[/%-]{2,}E)>? *');
            strList := TStringList.Create;
            enableDoc := true;
            try
                while not eof(inputFile) do
                begin
                    ReadLn(inputFile, line);
                    if reAll.Exec(line) then
                    begin
                        if enableDoc then
                        begin
                            if reEnd.Exec(line) then enableDoc := false
                            else if reDoc.Exec(line) then
                            begin
                                strList.Clear;
                                reDoc.Split(line, strList);
                                WriteLn(outputFile, strList.Strings[1]);
                            end
                        end
                        else if reBegin.Exec(line) then enableDoc := true;
                    end;
                end;
                Flush(outputFile);
            except
                on E: EInOutError do
                    ShowMessage( 'File Error: '+ E.ClassName + #10 + E.Message );
            end;
        finally
            if strList <> nil then strList.Free;
            if reAll <> nil then reAll.Free;
            if reDoc <> nil then reDoc.Free;
            if reBegin <> nil then reBegin.Free;
            if reEnd <> nil then reEnd.Free;
            if reValidFilename <> nil then reValidFilename.Free;
            CloseFile(inputFile);
            CloseFile(outputFile);
        end;
    end
    else
        Application.MessageBox ('Source file does not exist.', 'Error', mb_Ok or mb_IconHand);
    Sleep(1000);
    lblStatus.Caption := 'Finished, choose source and destine directory+filename.';
end;

//D>
//D> <a id="OnClick"></a>
//D> ### OnClick
//D>
//D> | Return  | Name        | Parameters           | Result |
//D> |  ------ | ----------- | -------------------- | ------ |
//D> | _None_  | **OnClick** | _TObject_ **Sender** | _None_ |
//D>
//D> Treat edit destine click event.
//D>
//D> **Parameter**:
//D> _TObject_ **Sender**: Objects and Events from main window.
//D>
//D> **Result**:
//D> _TFormMain_: Prepare destine edit text to show path + filename.
//D>
//D> [⬅️](btnRunClick) | [⬆️](#Index)

procedure TFormMain. OnClick( Sender: TObject);
var
    strFilename : TStringList;
begin
    try
        strFilename := TStringList.Create;
        strFilename.Clear;
        strFilename.Delimiter := '/';
        strFilename.StrictDelimiter := True;
        strFilename.DelimitedText := editSource.Text;
        if strFilename.Count > 0 then
        begin
            editDestine.FileName := strFilename.Strings[strFilename.Count - 1] + '.md';
            editDestine.Text := editDestine.InitialDir + '/' + editDestine.FileName;
        end;
    finally
        if strFilename <> nil then strFilename.Free;
    end;
end;

//D>
//D> <a id="OnCreate"></a>
//D> ### OnCreate
//D>
//D> | Return  | Name         | Parameters           | Result |
//D> |  ------ | ------------ | -------------------- | ------ |
//D> | _None_  | **OnCreate** | _TObject_ **Sender** | _None_ |
//D>
//D> Treat form create event.
//D>
//D> **Parameter**:
//D> _TObject_ **Sender**: Objects and Events from main window.
//D>
//D> **Result**:
//D> _TFormMain_: Prepare main form edit box to load and show current directory.
//D>
//D> [⬅️](OnClick) | [⬆️](#Index)

procedure TFormMain. OnCreate( Sender: TObject);
begin
    currentDir := GetCurrentDir;
    if DirectoryExists(currentDir + '/doc') then
        docDir := currentDir + '/doc'
    else
        docDir := currentDir;
    editSource.InitialDir := currentDir;
    editDestine.InitialDir := docDir;
end;

end.

File: makeDoc.py
#!/usr/bin/env python3
# coding=utf-8
##H>
##D><h1 id="top">makeDoc.py</h1>
import os
import re
import sys
import argparse

from modules import debug

##D><table id="description">
##D><tr><td align=right></td><td></td><td>Python program to export documentation from source code files.</td></tr>
##D><tr><td align=right><b>File</b></td><td>:</td><td>makeDoc.py</td></tr>
##D><tr><td align=right><b>Date</b></td><td>:</td><td>2025-09-30</td></tr>
##D><tr><td align=right><b>Author</b></td><td>:</td><td><a href="mailto:leandrohuff@email.com">Leandro</a></td></tr>
##D><tr><td align=right><b>Version</b></td><td>:</td><td>1.0.0 / 2025-09-30</td></tr>
##D><tr><td align=right><b>Copyright</b></td><td>:</td><td>CC01 1.0 Universal</td></tr>
##D></table>
##D><br>
##D><p>
##D><b>Note</b>: Changes in this documentation will be discarded on next build, any changes should be made on source code documentation instead. <br>
##D></p>


##D><br>
##D><h3 id="main">main( )</h3>
##D><i>integer</i> <b>main</b>( <i>string</i> <b>args</b> ) : <i>none</i> <br>
##D>&ensp;Main application program. <br>
##D><br>
##D><b>Parameter</b>: <br>
##D>&ensp;<i>string</i> <b>args</b> : Parameters from command line. <br>
##D>&ensp;<b>-h</b>, <b>--help</b>           show this help message and exit.  <br>
##D>&ensp;<b>-i</b>, <b>--input</b>  INPUT   Path and filename to the input file.  <br>
##D>&ensp;<b>-o</b>, <b>--output</b> OUTPUT  Path and filename to the output file.  <br>
##D>&ensp;<b>-f</b>, <b>--force</b>          Force yes for any question.  <br>
##D>&ensp;<b>-g</b>, <b>--debug</b>          Enable debug messages.  <br>
##D><br>
##D><b>Result</b>: <br>
##D>&ensp;<i>none</i> <br>
##D><br>
##D><b>Return</b>: <br>
##D>&ensp;<i>integer</i>: <b>0</b> - Success <br>
##D>&ensp;<i>integer</i>: <b>1</b> - Failure <br>
def main(args) -> int:
    errorCode: int = 0
    inputFilename: str = ""
    outputFilename: str = ""
    flagSaveFooter: bool = False
    flagMakeDoc : bool = True

    parser = argparse.ArgumentParser(
        description="Read taged lines documentation from source code and export documentation to output file."
    )

    parser.add_argument(
        "-i", "--input", help="Path and filename to the input file."
    )
    parser.add_argument(
        "-o", "--output", help="Path and filename to the output file."
    )
    parser.add_argument(
        "-f", "--force", action="store_true", help="Force yes for any question."
    )
    parser.add_argument(
        "-g", "--debug", action="store_true", help="Enable debug messages."
    )

    args = parser.parse_args()

    if args.debug:
        debug.setDebugFlag(True)
        debug.log("Debug enabled.")

    if args.force:
        debug.log("Force mode is enabled.")

    if not args.input:
        debug.error(f"Empty parameter for input file.")
        errorCode = 1
        return errorCode
    elif not os.path.exists(args.input):
        debug.info(f"Input file {args.input} not found, no documentaion exported.")
        errorCode = 2
        return errorCode

    inputFilename = args.input
    debug.log(f"Input file: {inputFilename}")

    if not args.output:
        debug.error(f"Empty parameter for output file.")
        errorCode = 3
        return errorCode
    elif not args.force and os.path.exists(args.output):
        debug.warning(
            f"Output file {args.output} already exist, no documentation exported."
        )
        errorCode = 4
        return errorCode

    outputFilename = args.output
    debug.log(f"Output file: {outputFilename}")

    padraoRegex: str = r"^ *([#|']+|[/|'|%|-]{2,}[B|D|E])>? *"
    padraoBeginDoc: str = r"^ *([#|']+|[/|'|%|-]{2,}B>?) *"
    padraoDoc: str = r"^ *([#|']+|[/|'|%|-]{2,}D>?) *"
    padraoEndDoc: str = r"^ *([#|']+|[/|'|%|-]{2,}E>?) *"

    try:
        with (
            open(inputFilename, "r") as inputFile,
            open(outputFilename, "w") as outputFile,
        ):
            # make Documentation
            for linhaFile in inputFile:
                matchDoc = re.search(padraoRegex, linhaFile)
                if matchDoc:
                    if flagMakeDoc and re.search(padraoDoc, linhaFile):
                        textDoc = re.split(padraoDoc, linhaFile)
                        outputFile.write(textDoc[2])
                    elif re.search(padraoEndDoc, linhaFile):
                        flagMakeDoc = False
                        debug.info ("Make Doc disabled.")
                    elif re.search(padraoBeginDoc, linhaFile):
                        flagMakeDoc = True
                        debug.info ("Make Doc enabled.")
    except FileNotFoundError as err:
        errorCode = 3
        debug.log(f"File not found: {err}")
    except IOError as err:
        errorCode = 4
        debug.log(f"File I/O error: {err}")
    return errorCode


##D><br>
##D><h3 id="entry">Python Entry Point</h3>
##D>&ensp;Main application entry point. <br>
##D><br>
##D><b>Parameter</b>: <br>
##D>&ensp;<i>string</i> <b>args</b> : Parameters from command line. <br>
##D><br>
##D><b>Result</b>: <br>
##D>&ensp;Call main() function to run all application. <br>
##D>&ensp;Print a success or failure message. <br>
##D><br>
##D><b>Return</b>: <br>
##D>&ensp;<i>integer</i>: <b>0</b> - Success <br>
##D>&ensp;<i>integer</i>: <b>1</b> - Failure <br>
##F>
if __name__ == "__main__":
    ret: int = main(sys.argv)
    if ret == 0:
        debug.info("Export documentation.")
    else:
        debug.error("Export documentation.")
    exit(ret)
